### PR TITLE
tests: make log monitor as common helper (followup #15667

### PR DESCRIPTION
**Pull request by fuweid**
It's followup of #15667 (tests: deflake TestV3WatchRestoreSnapshotUnsync).

This patch is to use zaptest/observer as base to provide a similar
function to pkg/expect.Expect.

The test env

```bash
    11th Gen Intel(R) Core(TM) i5-1135G7 @ 2.40GHz
    mkdir /sys/fs/cgroup/etcd-followup-15667
    echo 0-2 | tee /sys/fs/cgroup/etcd-followup-15667/cpuset.cpus # three cores
```

Before change:

* memory.peak: ~ 681 MiB
* Elapsed (wall clock) time (h:mm:ss or m:ss): 6:14.04

After change:

* memory.peak: ~ 671 MiB
* Elapsed (wall clock) time (h:mm:ss or m:ss): 6:13.07

Based on the test result, I think it's safe to be enabled by default.

Please read https://github.com/etcd-io/etcd/blob/main/CONTRIBUTING.md#contribution-flow.

### CODE REVIEW

File: tests/framework/testutils/log_observer.go [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+	"strings"
+	"sync"
+	"time"
+
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+	zapobserver "go.uber.org/zap/zaptest/observer"
+)
+
+type LogObserver struct {
```
Comment (serathius) [inline tests/framework/testutils/log_observer.go:29]: Please unify interface naming with `pkg/expect` as you seem to be inspired to it, but introduce your own naming scheme.

File: tests/framework/integration/cluster.go [hunk 2]
```diff
@@ -174,6 +174,8 @@ type ClusterConfig struct {
 	ExperimentalMaxLearners     int
 	DisableStrictReconfigCheck  bool
 	CorruptCheckTime            time.Duration
+
+	EnableMemberLogObserver bool
```
Comment (serathius) [inline tests/framework/integration/cluster.go:178]: Don't see any reson this cannot be enabled by default like in e2e package.

File: tests/framework/testutils/log_observer.go (hunk 1 above)
Comment (fuweid) [reply tests/framework/testutils/log_observer.go:29]: Hi @serathius , in pkg/expect package, it provides the `Expect(s string)` to return first expected line. However, we need to get first-N lines instead of one line. For instance, in `TestV3WatchRestoreSnapshotUnsync` test case, we need to know the leader has compacted twice. It's hard to follow the existing `pkg/expect` interface pattern.

**Comment by ahrtr** [comment]
Please mark this PR as ready to review once you think it's ready. Than I will take a look, thx

File: tests/framework/integration/cluster.go (hunk 2 above)
Comment (fuweid) [reply tests/framework/integration/cluster.go]: Updated!

File: tests/framework/testutils/log_observer.go [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+	for {
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		default:
+		}
+
+		entries, total := logOb.syncLogs()
+
+		for ; i < total; i++ {
```
[the remark this answers is not in the document]
Comment (fuweid) [reply tests/framework/testutils/log_observer.go:67]: Hi, `i` is increasing item and we don't need to scan the entities which has been scanned. So I declare the `i` before entering `loop`.

**Comment by ahrtr** [comment]
Currently there are two sets of log monitoring, one is `pkg/expect` which is only used by e2e test cases. The other is `LogObserver` implemented in this PR, which is only used by integration test cases. They have different log sources, `pkg/expect` reads logs from process/command output, while `LogObserver` duplicates the log from each integration member. I think it would be better if we can unify them into one log monitoring utility.  Please see rough design below,

log_monitor

File: tests/framework/testutils/log_observer.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (ahrtr) [reply tests/framework/testutils/log_observer.go:67]: Please see my comment above `Note the ob.TakeAll() (line 101) truncates all observed logs.`. It means each time `logOb.syncLogs()` (line 65) will get different log slice, so my understanding is that you should reset i to 0.  Does it make sense?

File: tests/framework/integration/cluster.go [hunk 4]
```diff
@@ -743,14 +747,23 @@ func MustNewMember(t testutil.TB, mcfg MemberConfig) *Member {
...
+
+	options := zaptest.WrapOptions(
+		zap.Fields(zap.String("member", name)),
+
+		// copy logged entities to log observer
+		zap.WrapCore(func(oldCore zapcore.Core) zapcore.Core {
+			return zapcore.NewTee(oldCore, obCore)
+		}),
+	)
+	return zaptest.NewLogger(t, zaptest.Level(level), options).Named(name), logOb
```
[the remark this answers is not in the document]
Comment (fuweid) [reply tests/framework/integration/cluster.go:766]: Yeah. First version is to disable observer by default. I was concerned about memory usage because the github action node nanny might send the sigkill randomly. I think it can be good to be optional item.

File: tests/framework/testutils/log_observer.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (fuweid) [reply tests/framework/testutils/log_observer.go:67]: `syncLogs` is to dump the logged entries from `zapobserver` and truncates all the entries in `zapobserver` side. It's like `move` action.

For example, 

| Sync Time  |  LogObserver entries  |  zapobserver entries |
| ----- | --- | --- |
|  | [] | [A, B ] |
| T1 | [A, B] | [] |
|  | [A, B] | [C, D, F] |
| T2 | [A, B, C, D, F] | [] |

The `LogObserver` always contains all logged entries after synced. Even if the `entries`'s underlaying slice(memory address) has been changed by go-runtime `append`, both the order and the content are the same. So I think we can reuse the `i` to continue expect action. Is it acceptable?

[the remark this answers is not in the document]
Comment (ahrtr) [reply tests/framework/testutils/log_observer.go:67]: Got it. But I don't understand why you cache all entries? Can we change line 101 to something like below?

```
return logOb.ob.TakeAll()
```

Also why do you return the two values in `(*LogObserver) syncLogs()`, users can easily get the count.

[the remark this answers is not in the document]
Comment (ahrtr) [reply tests/framework/testutils/log_observer.go:67]: It seems that If we monitor/analyze the log multiple times, then we still need to cache all log entries.

[the remark this answers is not in the document]
Comment (fuweid) [reply tests/framework/testutils/log_observer.go:67]: > Also why do you return two values in (*LogObserver) syncLogs(), users can easily get the count.

`append` might grow the slice's capacity and then next `append` might not copy. Since I don't use lock when I scan the entries, I think it's better to should return slice with len so that there is no race issue. I will add a comment to the function.

> It seems that If we monitor/analyze the log multiple times, then we still need to cache all log entries.

Yes. It allows caller to expect multiple times. It is aligned with `pkg/expect` behavior.

[the remark this answers is not in the document]
Comment (ahrtr) [reply tests/framework/testutils/log_observer.go:67]: > Since I don't use lock when I scan the entries, I think it's better to should return slice with len so that there is no race issue.

It isn't correct to me. No matter how you update/append the `logOb.entries`, it will not affect the already returned slice(entries at line 65). Note that in golang, a slice is actually a struct (with 3 fields), it's returned by value.

[the remark this answers is not in the document]
Comment (fuweid) [reply tests/framework/testutils/log_observer.go:67]: You're correct. The slice header maintains the `Len` even if they shares the same underlaying `Data`. So it's fine to return slice only. I will update it. Thanks for pointing it out.

**Comment by fuweid** [comment]
@ahrtr Thanks for sharing the idea about unify-log-expect. I need to rethink this part and seek a way to build redirect-logger module for this.

File: tests/framework/testutils/log_observer.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (fuweid) [reply tests/framework/testutils/log_observer.go]: updated! PTAL, thanks

File: tests/framework/integration/cluster.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (fuweid) [reply tests/framework/integration/cluster.go:766]: I retest it and it doesn't bring too many overhead. So I keep it as it is.

File: tests/framework/testutils/log_observer.go [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+		}
+
+		entries := logOb.syncLogs()
+
+		// The order of entries won't be changed because of append-only.
+		// It's safe to skip scanned entries by reusing `i`.
+		for ; i < len(entries); i++ {
+			buf, err := logOb.enc.EncodeEntry(entries[i].Entry, entries[i].Context)
+			if err != nil {
+				return nil, fmt.Errorf("failed to encode entry: %v", err)
```
[the remark this answers is not in the document]
Comment (fuweid) [reply tests/framework/testutils/log_observer.go]: Updated!

File: tests/framework/testutils/log_observer_test.go [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+func TestLogObserver_Timeout(t *testing.T) {
+	logCore, logOb := NewLogObserver(zap.InfoLevel)
+
+	logger := zap.New(logCore)
+	logger.Info(t.Name())
+
+	ctx, cancel := context.WithTimeout(context.TODO(), 100*time.Millisecond)
+	_, err := logOb.Expect(ctx, "unknown", 1)
+	cancel()
+	assert.Equal(t, true, errors.Is(err, context.DeadlineExceeded))
```
[the remark this answers is not in the document]
Comment (fuweid) [reply tests/framework/testutils/log_observer_test.go]: Looks good. Updated!

**Review by serathius — APPROVED** [review]
